nav2_mppi_controller/test/critics_tests.cpp

```diff
@@ -242,6 +242,86 @@ TEST(CriticTests, GoalAngleCritic)
   EXPECT_NEAR(costs(0), 9.42, 0.02);  // (3.14 - 0.0) * 3.0 weight
 }
 
+TEST(CriticTests, GoalAngleCriticSymmetricYawTolerance)
+{
+  // Standard preamble
+  auto node = std::make_shared<rclcpp_lifecycle::LifecycleNode>("my_node");
+  auto costmap_ros = std::make_shared<nav2_costmap_2d::Costmap2DROS>(
+    "dummy_costmap", "", "dummy_costmap");
+  ParametersHandler param_handler(node);
+  rclcpp_lifecycle::State lstate;
+  costmap_ros->on_configure(lstate);
+
+  // Set parameters for critic with symmetric disabled
+  node->declare_parameter("critic_sym_disabled.cost_power", 1);
+  node->declare_parameter("critic_sym_disabled.cost_weight", 3.0);
+  node->declare_parameter("critic_sym_disabled.threshold_to_consider", 0.5);
+  node->declare_parameter("critic_sym_disabled.symmetric_yaw_tolerance", false);
+
+  // Set parameters for critic with symmetric enabled
+  node->declare_parameter("critic_sym_enabled.cost_power", 1);
+  node->declare_parameter("critic_sym_enabled.cost_weight", 3.0);
+  node->declare_parameter("critic_sym_enabled.threshold_to_consider", 0.5);
+  node->declare_parameter("critic_sym_enabled.symmetric_yaw_tolerance", true);
+
+  models::State state;
+  models::ControlSequence control_sequence;
+  models::Trajectories generated_trajectories;
+  generated_trajectories.reset(1000, 30);
+  models::Path path;
+  xt::xtensor<float, 1> costs = xt::zeros<float>({1000});
+  float model_dt = 0.1;
+
+  // Setup goal angle critic with symmetric_yaw_tolerance disabled (default)
+  GoalAngleCritic critic;
+  critic.on_configure(node, "mppi", "critic_sym_disabled", costmap_ros, &param_handler);
+  EXPECT_EQ(critic.getName(), "critic_sym_disabled");
+
+  // Setup goal angle critic with symmetric_yaw_tolerance enabled
+  GoalAngleCritic critic_symmetric;
+  critic_symmetric.on_configure(node, "mppi", "critic_sym_enabled", costmap_ros, &param_handler);
+  EXPECT_EQ(critic_symmetric.getName(), "critic_sym_enabled");
+
+  // Setup state and path: robot at origin, goal at (10, 0) with yaw of PI
+  state.pose.pose.position.x = 9.7;  // Within threshold_to_consider (0.5) of goal at 10.0
+  path.reset(10);
+  path.x(9) = 10.0;
+  path.y(9) = 0.0;
+  path.yaws(9) = 3.14159;  // Goal at PI
+
+  // Test 1: Trajectory at 0 rad (forward) with symmetric disabled
+  // Should have high cost since it's not aligned with goal at PI
+  generated_trajectories.yaws = xt::zeros<float>({1000, 30});
+  costs = xt::zeros<float>({1000});
+  CriticData data1 = {state, generated_trajectories, path, costs, model_dt, false, nullptr, nullptr, std::nullopt, std::nullopt};
+  data1.motion_model = std::make_shared<DiffDriveMotionModel>();
+  critic.score(data1);
+  float cost_forward_without_symmetric = costs(0);
+  EXPECT_GT(cost_forward_without_symmetric, 0.1);  // Should have significant cost
```

Comment:
For all tests: calculate the cost by hand so you know what it is and check its within a tight tolerance of that -- not simply non-zero.

Reply:
@sandeepdutta see this too